### PR TITLE
Events deduplication improvements

### PR DESCRIPTION
**Pull request by monsha**
This PR adds a new model (DeduplicationToken) where a Digest :: SHA256 will be saved for each event payload, which allows comparing the sha string instead of the whole payload.

Add a new rake task (deduplication:remove_old_tokens) to keep clean the table, removing all tokens with more than three months created.

Add a migration to preload the tokens corresponding to the last events of the agents, so that the events are not duplicated.

Deploy first this: https://github.com/chattermill/huginn/pull/106

### CODE REVIEW

File: app/concerns/deduplication_concern.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  # This method returns true if the result should be stored as a new event.
+  # If mode is set to 'on_change', this method may return false and update an existing
+  # event to expire further in the future.
+  def store_payload?(old_events, payload)
+    case interpolated['mode'].presence
+    when 'on_change'
+      token = payload_to_sha(payload)
+      if found = old_events.find { |e| e.token == token }
```
Comment (OpakAlex) [inline app/concerns/deduplication_concern.rb:27]: maybe remove if?)
`!old_events.find { |e| e.token == token }`

**Comment by monsha** [comment]
@OpakAlex Could you please review this? Take a look at the migration of tokens, the goal is to preload the tokens corresponding to the latest agents events, so that when we implement this new way of looking for duplicates there is data and the events are not duplicated. I tested it locally and It last around 8secs.

File: app/concerns/deduplication_concern.rb (hunk 1 above)
Comment (monsha) [reply app/concerns/deduplication_concern.rb:27]: @OpakAlex the "If" is necessary because in case that found it, It need to update the event

**Review by OpakAlex — APPROVED** [review]

**Comment by duboff** [comment]
@monsha can we run some sort of benchmark to compare how much performance gain this generates vs. status quo?

**Comment by monsha** [comment]
@duboff: 

# Comparing simple hash vs string:
```
payloads = []
tokens = []
1000000.times do
  p = {"key": rand(100)}
  payloads.push(p)
  tokens.push(Digest::SHA256.hexdigest(p.to_json))
end

payload = payloads.last
token = Digest::SHA256.hexdigest(payload.to_json)

Benchmark.bm do |x|
  x.report(:payload) { payloads.find { |p| p == payload } }
  x.report(:token) { tokens.find { |t| t == token } }
end
```
```
                user     system      total        real
payload  0.000000   0.000000   0.000000 (  0.000078)
token      0.000000   0.000000   0.000000 (  0.000029)
```
* more than 50% faster


# whole payload (hash) vs token (string)
```
a = Agent.last
events = a.events.limit(1000)
tokens = a.tokens.limit(1000)

payload = {"some key"=>"not existing"}
token = Digest::SHA256.hexdigest(payload.to_json)
```

### 1 new event (uniqueness_look_back of 1000)
```
Benchmark.bm do |x|
  x.report(:payload) { events.find { |e| e.payload == payload } }
  x.report(:token) { tokens.find { |e| e.token == token }  }
end
```
```
                user     system      total        real
payload  0.020000   0.000000   0.020000 (  0.021205)
token      0.000000   0.000000   0.000000 (  0.005808)
```

### 1000 new events (uniqueness_look_back of 1000)
```
Benchmark.bm do |x|
  x.report(:payload) { 1000.times do ;events.find { |e| e.payload == payload }; end }
  x.report(:token) { 1000.times do ; tokens.find { |e| e.token == token }; end }
end
```

```
                user     system      total        real
payload  0.840000   0.020000   0.860000 (  0.867093)
token      0.550000   0.010000   0.560000 (  0.565537)
```

**Comment by duboff** [comment]
💯 Niice

**Review by duboff — APPROVED** [review]
LGTM

@monsha Pls deploy but monitor huggin for an hour or so after to make sure nothing weird is happening

**Comment by monsha** [comment]
@duboff I'm going to split this PR so there is no downtime or errors, because at the time this is implemented the agents will look inside the table of tokens to see if there are duplicates and at the beginning the table will be empty (until `rake db: migrate` is not executed) and all events will be added.

**Pull request merged**